### PR TITLE
Passivate imported keys if the associate certificate is expired

### LINKED ISSUE #34973: ES256 key continue to be used to sign token even after expiry
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

oidc

### Describe the bug

I created a java key provider using an ES256 key in the keystore. It works fine and issues access tokens signed by this key.
But even after the key has expired, keycloak continues to use it to sign tokens.
I would have expected, a default ES256 key to be generated and used instead of the expired key.
May be it works like so because KeyCloak does not support ES256 default keys (only RSA256?)

### Version

26.0.0

### Regression

- [ ] The issue is a regression

### Expected behavior

The expired ES256 key should not be used to issue tokens. A new default ES256 key should be generated and used instead or a RSA256 token be issued etc.

### Actual behavior

expired ES256 key continue to be used to sign access tokens.

### How to Reproduce?

Create a java keystore that contains an ES256 key. Make the key expire in 1 day for example.
Use that to create a java key provider and make this a default for a realm.
Wait for the key to expire and login to the realm.
You will see the expired kid still being used.

### Anything else?

image
expired key
issued token:
eyJhbGciOiJFUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJOU2pZeF9jRlFCekxUUV9EUnVsWVNxb0Jzdy10b0lYdHU0WEJPRHFGUGhBIn0.eyJleHAiOjE3MzE2MTc5ODMsImlhdCI6MTczMTYxNzY4MywiYXV0aF90aW1lIjoxNzMxNjE3NjgzLCJqdGkiOiJhYWFlY2VkNy00MjdmLTQ5NWYtOTAxZi1kYmI5MGI2NjdmY2YiLCJpc3MiOiJodHRwOi8vbG9jYWxob3N0OjgwODAvcmVhbG1zL2ludGVncmF0aW9uIiwiYXVkIjoiYWNjb3VudCIsInN1YiI6ImE5NWY1ODQ3LWNiMmEtNDA1Yy1hMjYzLWNlZWFiMGQ0NGE4OSIsInR5cCI6IkJlYXJlciIsImF6cCI6ImFjY291bnQtY29uc29sZSIsInNpZCI6ImI0YTgxM2EwLTBjNTMtNGFkNS04NTExLTM2ZWFlMGIyZDM1MSIsImFjciI6IjEiLCJyZXNvdXJjZV9hY2Nlc3MiOnsiYWNjb3VudCI6eyJyb2xlcyI6WyJtYW5hZ2UtYWNjb3VudCIsIm1hbmFnZS1hY2NvdW50LWxpbmtzIl19fSwic2NvcGUiOiJvcGVuaWQgZW1haWwgcHJvZmlsZSIsImVtYWlsX3ZlcmlmaWVkIjpmYWxzZSwibmFtZSI6Ik5hdiBLdW1hciIsInByZWZlcnJlZF91c2VybmFtZSI6Im5hdnRlc3R1c2VyIiwiZ2l2ZW5fbmFtZSI6Ik5hdiIsImZhbWlseV9uYW1lIjoiS3VtYXIiLCJlbWFpbCI6Im5rdW1hcjE5QGxlbm92by5jb20ifQ.YtmWHxUaAqXqIuKn9E4uyJOE9k899u4haNKdfW7vEAvQ_3oCWPPhNQn8cZbdmUBHnMPmTO0vjTR7jzeoWNX7nw
===
{
  "alg": "ES256",
  "typ": "JWT",
  "kid": "NSjYx_cFQBzLTQ_DRulYSqoBsw-toIXtu4XBODqFPhA"
}
===
decoded JWT details:
{
  "exp": 1731617983,
  "iat": 1731617683,
  "auth_time": 1731617683,
  "jti": "aaaeced7-427f-495f-901f-dbb90b667fcf",
  "iss": "http://localhost:8080/realms/integration",
  "aud": "account",
  "sub": "a95f5847-cb2a-405c-a263-ceeab0d44a89",
  "typ": "Bearer",
  "azp": "account-console",
  "sid": "b4a813a0-0c53-4ad5-8511-36eae0b2d351",
  "acr": "1",
  "resource_access": {
    "account": {
      "roles": [
        "manage-account",
        "manage-account-links"
      ]
    }
  },
  "scope": "openid email profile",
  "email_verified": false,
  "name": "Nav Kumar",
  "preferred_username": "navtestuser",
  "given_name": "Nav",
  "family_name": "Kumar",
  "email": "nkumar19@lenovo.com"
}

### PR DESCRIPTION
**Pull request by rmartinc**
Closes #34973

This PR passivate (`active=false`) an imported key if the associated certificate is expired. It only works for imported certificates (`java-keystore` and `rsa` provider types) and not for the generated providers. The generated keys sets the validity of their certificates (if present) to 10 years, very long time, so not doing anything there. Now, when the key/certificate is read and cached in the model, the expiration date is also added as another note and the expiration is checked in any use.

Another change is the `java-keystore` has been modified to validate the certificate chain only at import. Previously it was checked when loading and that made the key disappear if expired. That behavior can affect valid issued tokens. Now it follows the same idea, the key is passivated, so it's not active for new tokens but still present for previous ones.

The `verifiable-credentials-realm.json` needs to remove the keys because they are expired and it's a imported rsa key. It's just a leftover so I just removed that part to create new keys at realm import time.

New tests added for the providers. For them I needed to create certificates with validity of one hour. So I needed to add a new variant to pass the expiration date in the certificate creation methods. 

Do we add some notes here? In the upgrading section? And, if yes, to what version 26.1.0 or a backport to the next 26.0.x?

### CODE REVIEW

File: services/src/main/java/org/keycloak/keys/AbstractImportedRsaKeyProviderFactory.java [hunk 1]
```diff
@@ -81,9 +82,15 @@ public void validateConfiguration(KeycloakSession session, RealmModel realm, Com
             if (!certificate.getPublicKey().equals(keyPair.getPublic())) {
                 throw new ComponentValidationException("Certificate does not match private key");
             }
+
+            try {
+                certificate.checkValidity();
+            } catch (CertificateException e) {
+                throw new ComponentValidationException("Certificate is not valid");
```
Comment (mposolda) [inline services/src/main/java/org/keycloak/keys/AbstractImportedRsaKeyProviderFactory.java:89]:
```suggestion
                throw new ComponentValidationException("Certificate is not valid", e);
```

Minor: Is it possible to keep the original exception and make sure that when administrator tries to upload the key with expired certificate, the underlying cause is logged as well - at least to server.log, but maybe even to the error message itself if possible, so it is returned in the response? At least, that is what I usually preferred when using things as the user to have more details if possible if something doesn't work :-)

**Review by mposolda — COMMENTED** [review]
@rmartinc Thanks! +1 for add something to the upgrading documentation as it is possible that people have expired certificates in their realm JSON configs and things may stop working on them (as can be seen in our testsuite).

**Review by edewit — DISMISSED** [review]
it's fine from a ui perspective

File: services/src/main/java/org/keycloak/keys/AbstractImportedRsaKeyProviderFactory.java (hunk 1 above)
Comment (rmartinc) [reply services/src/main/java/org/keycloak/keys/AbstractImportedRsaKeyProviderFactory.java]: :+1: Yes, it's much better. Done now!

**Comment by rmartinc** [comment]
@mposolda Added the upgrading note and your suggestion with the exception.

**Review by mposolda — APPROVED** [review]
@rmartinc Thanks!

**Pull request merged**